lib/spack/spack/variant.py

```diff
@@ -852,6 +853,27 @@ def disjoint_sets(*sets):
     return DisjointSetsOfValues(*sets).allow_empty_set().with_default('none')
 
 
+@functools.total_ordering
+class Value(object):
```

Comment:
Minor nitpick: everything else in the DSL is lowercase. Shouldn’t this be too? We have `patch()` which can be nested in a `depends_on()`, which is similar to this, and it’s lowercase.

Reply:
In class definitions, though?  These are all executed before any functions in the package are called, and anything in the functions would be local.

Reply:
Ok, running through all `builtin` packages:

h2. Class scope
This happens in only two places at `class` scope, due to the loops in `CudaPackage`:

```python
    # Make sure cuda_arch can not be used without +cuda
    for value in cuda_arch_values:
        conflicts('~cuda', when='cuda_arch=' + value)
```

 and `RocmPackage`:

```python
    # Make sure amdgpu_targets cannot be used without +rocm
    for value in amdgpu_targets:
        conflicts('~rocm', when='amdgpu_target=' + value)
```

Subclasses of these end up with `value` set to either `86` or `"gfx1012"` -- the last values in `cuda_arch_values`
 and `rocm_arch_values`, respectively.  But that actually won't interfere with `value()` in subclasses -- superclass attributes don't end up in global scope for subclass definitions.  And in both of these classes, the for loop is the last or nearly the last thing in the class definition, so the loop variable wouldn't interfere with `value()` directives in any `variant` definitions.  *Most* packages have any iteration logic like this after all the directives.

I also went through all `builtin` packages and checked whether `value` was globally defined anywhere.  It's not.

So IMO this is pretty safe as a new directive for use within `variant()`.

Reply:
I checked `CudaPackage` and `RocmPackage` and I think we can safely remove the loops in question, since the architecture-related variants are now conditional (see https://github.com/spack/spack/pull/29530/commits/150dd8d326f3dab1d5d3523ce2398631b55a6102)